File: 프로그래머스/lv1/42576. 완주하지 못한 선수/완주하지 못한 선수.py
# def solution(participant, completion):
#     participant.sort()
#     completion.sort()

#     for i, z in zip(participant, completion):
#         if i != z:
#             return i
#     return participant[-1]

# from collections import Counter

# def solution(participant, completion):
#     return ''.join(Counter(participant) - Counter(completion))


def solution(participant, completion):
    d = {}
    for i in participant:
        d[i] = d.get(i, 0) + 1
    
    for i in completion:
        d[i] -= 1 
        
    dnf = [k for k, v in d.items() if v > 0]
    
    return dnf[0]
    

File: 프로그래머스/lv1/42576. 완주하지 못한 선수/완주하지 못한 선수.java
// import java.util.*;

// class Solution {
//     public String solution(String[] participant, String[] completion) {
//         String answer = "";

//         Map<String, Integer> map = new HashMap<>();
//         for (int i = 0; i < completion.length; i++) {
//             map.put(completion[i], 1);
//         }

//         for (int i = 0; i < completion.length; i++) {
//             if (!(map.containsKey(participant[i]))) {
//                 answer = participant[i];
//             }
//         }
//         return answer;
//     }
// }

import java.util.*;

class Solution {
    public String solution(String[] participant, String[] completion) {
        String answer = "";

        Map<String, Integer> map = new HashMap<>();
        for(String name: participant) map.put(name, map.getOrDefault(name, 0) + 1);

        for(String name: completion) {
            map.put(name, map.get(name) - 1);
        }

        for(String key: map.keySet()) {
            if (map.get(key) != 0) {
                answer = key;
            }
        }

        return answer;
 }
}
